feat: add support for controlling bundle size

techfg opened the pull request:
As discussed in #326, other than for `astro-expressive-code`, there is no current way to control the bundle size when using `expressive-code` outside of some `package.json` related workarounds.  This PR aims to address that by providing stock support for controlling bundle size by introducing a new `Core` concept across the various packages (e.g., `ExpressiveCodeCore`, `RehypeExpressiveCodeCore`, etc.).

> [!NOTE]
> The approach taken in this PR is to ensure 100% backwards compatibility with v0.41.2.  All existing tests pass without any modification.  If it's acceptable to introduce some minor adjustments, the current approach can be simplified in some areas. Since `expressive-code` uses semver `0.x.x`, technically every release is a breaking change so for the purposes of long-term maintenance, some changes may be worth considering.

## TLDR;

| 0 lang / 0 theme                                | Full Bundle | Shiki     | PluginShiki | PluginFrames | PluginTextMarkers |
| ----------------------------------- | ----------- | --------- | ----------- | ---------------- | --------------------- |
| v0.41.2                           | 9,577,922   | 9,695,299 | 20,365      | 23,186           | 22,246                |
| Core No Plugins                     | 347,138     | 0         | 0           | 3,788            | 2,367                 |
| Savings                             | 9,230,784   | 9,695,299 | 20,365      | 19,398           | 19,879                |

| 1 lang / 1 theme                                | Full Bundle | Shiki     | PluginShiki | PluginFrames | PluginTextMarkers |
| ----------------------------------- | ----------- | --------- | ----------- | ---------------- | --------------------- |
| v0.41.2                           | 9,577,922   | 9,695,299 | 20,365      | 23,186           | 22,246                |
| Core Bundle | 780,831     | 358,594   | 20,179      | 3,788            | 2,367                 |
| Savings                             | 8,797,091   | 9,336,705 | 186         | 19,398           | 19,879                |

| 1 lang / 1 theme                                | Full Bundle | Shiki     | PluginShiki | PluginFrames | PluginTextMarkers |
| ----------------------------------- | ----------- | --------- | ----------- | ---------------- | --------------------- |
| v0.41.2                           | 9,577,922   | 9,695,299 | 20,365      | 23,186           | 22,246                |
| Core Highlighter | 781,711     | 360,766   | 18,950      | 3,788            | 2,367                 |
| Savings                             | 8,796,211   | 9,334,533 | 1,415       | 19,398           | 19,879                |

1. Setting `shiki=false` in configuration makes no difference to final bundle size when using `ExpressiveCode/rehypeExpressiveCode/createRenderer`, to reduce bundle size, you must use `Core` directly
2. Shiki & Plugin* columns are `renderedLength` of corresponding item, the `Full Bundle` column is the actual final size of the module on disk
3. The reason `PluginFrames` and `PluginTextMarkers` are not zero in `Core` is primarily due to styles (css) which could likely be refactored out

## Additional Information
1. `remark-expressive-code` appears to be deprecated so it was not updated for `Core`
2. `astro-expressive-code` needs to be updated to `Core`. Once a decision is made on potentially incorporating this PR and the `Core` API finalized, I'll look at updating it to use `Core` rather than its `vite-plugin` approach.
3. I've added a few comments to code in this PR to highlight some things that should be specifically evaluated/considered outside of the general PR itself
4. The [bundle-size tests](packages/rehype-expressive-code/test/bundle-size.test.ts) generate builds across the different variants of `RehypeExpressiveCode` and `RehypeExpressiveCodeCore`.  In `dist` folder of each fixture are statistics files that can be reviewed for detailed information about the build output.  See [visualizer](https://github.com/expressive-code/expressive-code/compare/main...techfg:expressive-code:feat/support-controlling-bundle-size?expand=1#diff-c27836fbe666b134979131fe3f1fff2d576e70424548e8e1346a95b9cb764fccR96) and [bundle-stats](https://github.com/expressive-code/expressive-code/compare/main...techfg:expressive-code:feat/support-controlling-bundle-size?expand=1#diff-c27836fbe666b134979131fe3f1fff2d576e70424548e8e1346a95b9cb764fccR106) for more information.  

Open to any and all feedback, appreciate your consideration!

Resolves #326

techfg commented on packages/@expressive-code/plugin-shiki/src/index.ts:
```diff
@@ -25,7 +15,9 @@ export interface PluginShikiOptions {
 	 * @example { 'mjs': 'javascript' }
 	 */
-	langAlias?: Record<string, string> | undefined
+	// TODO: This should be simply L but for backwards compat, need to support string. This should be changed so that you cannot provide
```
I believe this is a bug in the current released version which uses `Record<string, string>`, however the `languages` (the value of the key) should be constrained to the languages available in the current bundle.  Currently, you could write an alias of `foo->bar` and would not get any typescript warning (it would fail at runtime).  Additionally, with `string, string`, you do not get any intellisense/autocomplete.  I've adjusted to use `StringLiteralUnion` which will solve the intellisense problem but I believe this should just be `L` in order to provide proper type safety.  I made this backwards compatible (will support any string) per the goals of the first draft of this PR but my recommendation is that this should be changed to L.  It wouldn't be considered a breaking change, I don't believe, because the current version will fail at runtime anyway.

hippotastic replied:
This was intentionally not limited to the bundled languages as users can add their own languages through the `langs` option, and it should be possible to define aliases for such additional languages as well. But using `StringLiteralUnion` is a great idea.

techfg replied:
Ah, I see what you did [here](https://github.com/expressive-code/expressive-code/blob/85a36414668fe60a418e0f6910a7c62373a972ef/packages/%40expressive-code/plugin-shiki/src/highlighter.ts#L105) now, nice :)  Sorry, overlooked the way this was handled previously.

Ok, I just committed https://github.com/expressive-code/expressive-code/pull/335/commits/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8 which makes the following adjustments:

1. Removes the [TODO](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9R21)
2. Adds a [note](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9R15) in the `langAlias` doc regarding `name` property to make it clear how it must align (`name` is required on [LanguageRegistration](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9R15) but this should help make it clear how `langAlias` matches up
3. Aligned the docs for [PluginShikiOptions](https://github.com/expressive-code/expressive-code/pull/335/files#diff-3a8ca5dc7c3fafd14174149f8941f7dc9d1dc90bb1cae3d1a3113a9579ba6166R8) and [PluginShikiBundleOptions](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9R58) regarding `langAlias`
4. Removed [langAlias](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9R96) from `PluginShikiHighlighterOptions` since it never really made sense to support it in the first place since the user is under full control of the highlighter and can specify `langAlias` directly
5. [Added tests](https://github.com/expressive-code/expressive-code/commit/58ec5b5ce1c5e046a35cfafb8420d47c06efc3e8#diff-9d6158838c242a688736b5129e5dd29d29689f2f44737e56dc2bc919e6e99103R516) regarding custom languages and aliases since I didn't see any coverage for this scenario in the existing tests.

techfg commented on packages/@expressive-code/plugin-shiki/src/index.ts:
```diff
@@ -271,3 +370,13 @@ function getRemovedRanges(original: string, edited: string): [start: number, end
+async function createRegexEngine(engine: PluginShikiOptions['engine']) {
+	// The [...engine...][0] syntax makes it easier to find this code in the built package,
+	// allowing astro-expressive-code to remove unused engines from the SSR bundle
+	// TODO: This could be adjusted to use direct imports if/when astro-expressive-code is updated
```
If `astro-expressive-code` is updated to use `Core`, these can become static imports vs. dynamic

techfg commented on packages/rehype-expressive-code/src/index.ts:
```diff
@@ -114,39 +165,72 @@ export type RehypeExpressiveCodeDocument = {
 }
 
+export type RehypeExpressiveCodeCoreRenderer = RehypeExpressiveCodeEngineRenderer<ExpressiveCodeCore>
+export type RehypeExpressiveCodeRenderer = RehypeExpressiveCodeEngineRenderer<ExpressiveCode>
```
These wrapper types only exists for backwards compat.  Ultimately, `ec` in the base type is the instance of `ExpressiveCode/ExpressiveCodeCore` that was created but from a typing perspective, I believe it makes more sense to just have `ec: ExpressiveCodeEngine` vs the actual instance types.  If I'm understanding the code, `ExpressiveCode` class is just a tiny layer on top of the engine.  If `ExpressiveCode/ExpressiveCodeCore` classes are thought of as a `bootstrap` for the engine, they shouldn't ever have any specific API so exposing the engine directly would simplify the typings here and in a couple of other places.

techfg replied:
Sounds good, will leave as-is, closing this as resolved.

techfg replied:
Further thoughts on this at https://github.com/expressive-code/expressive-code/pull/335#discussion_r2072862226

techfg commented on packages/rehype-expressive-code/src/index.ts:
```diff
@@ -114,39 +165,72 @@ export type RehypeExpressiveCodeDocument = {
+): Promise<RehypeExpressiveCodeEngineRenderer<E>> {
 	// Transfer deprecated `theme` option to `themes` without triggering the deprecation warning
-	const deprecatedOptions: Omit<RehypeExpressiveCodeOptions, 'theme'> & { theme?: ThemeObjectOrShikiThemeName | ThemeObjectOrShikiThemeName[] | undefined } = options
+	const deprecatedOptions: Omit<RehypeExpressiveCodeCommonOptions<ExpressiveCodeCoreConfig, E, T>, 'theme'> & {
```
Seems to be deprecated over a year ago, is it time to remove completely?  While this PR does not currently make any changes to the current public API, if we do make some minor ones, might be a good time to remove support for `theme`.  Also, since EC is `v0.x`, technically every release is to be considered a potential `breaking change` by package consumers :)

hippotastic replied:
Good question. I've just recently come across a user who still used the old `theme` option. Without this encounter, I would have probably agreed. :) I think I'll remove all deprecated features when finally switching to `v1.0`, which I'm planning to do this year. I'm not a fan of the special pre-`v1.0`-semver behavior and will be happy to get out of this zone. :)

techfg replied:
Sounds good, will leave as-is, closing this as resolved.

techfg commented on packages/rehype-expressive-code/test/bundle-size.test.ts:
```diff
@@ -0,0 +1,275 @@
+
+type ShikiDetails = Awaited<ReturnType<typeof buildFixture>>
+
+describe('integration with shiki affect on ec bundle size', () => {
```
This test suite adds ~25seconds to `pnpm test`.  I think there is a lot of value in having all these tests to inspect/review various configurations but for `normal` CI runs, possibly narrow this down to a subset of these across the 4 variation groups.  Each individual test takes ~1.5sec.

hippotastic replied:
It's awesome that you've added these tests and agree that there is a lot of value in having them.

In case we need to limit the tests performed during normal runs to prevent slowing down development too much, which subset would you recommend to always run?

techfg replied:
The more that I've thought about this, I think for CI, leaving all the tests makes sense - at least for the `main` branch workflow, possibly `pr` branches can have a subset.

Where this additional ~25secs being a challenge is in local development, possibly that is what you meant by "slowing down development"?  I think this question really comes down to what the normal workflow is for a developer when testing EC.  If the approach is using `pnpm test` from the root, then this additional time isn't ideal.

I think at minimum, the tests with `no-themes.ts` would be a good baseline across all 4 test categories within the suite.  This makes sure that a vanilla configuration with no additional themes specified works as expected re: bundle size for `ec`, `ec-core`, `ec-core-bundle` and `ec-core-highlighter`.

hippotastic commented:
This is seriously impressive work! Thank you for diving so deep into this, finding clever solutions, documenting your decisions, adding tests - just wow. I would have never expected this amount of effort being put into limiting the bundle size.

I've just returned from holidays and will have a look at this in more detail soon. Based on how substantial these changes are, going through them could take a while, but I'll make sure to make room for it in my schedule. Thanks again!

techfg commented:
> This is seriously impressive work! Thank you for diving so deep into this, finding clever solutions, documenting your decisions, adding tests - just wow. I would have never expected this amount of effort being put into limiting the bundle size.
> 
> I've just returned from holidays and will have a look at this in more detail soon. Based on how substantial these changes are, going through them could take a while, but I'll make sure to make room for it in my schedule. Thanks again!

Welcome back @hippotastic, hope you enjoyed the time away!  

Appreciate your kind words, looking forward to hearing your further thoughts once you've had a chance to fully digest.  In meantime, I'll start reviewing your feedback thus far and comment back and/or update PR accordingly.

techfg commented:
@hippotastic - 

I was able to figure out why `plugin-frames` and `plugin-text-markers` had a small footprint with the initial proposed `Core` layer introduced in this PR.  In short, both of these plugins, along with `expressive-code` itself have side-effects related to the EC styling system.  For example, [pluginFrameTexts](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-frames/src/index.ts#L75) prohibits the bundler from tree-shaking even when `plugin-frames` is not used since the `import { pluginFrames } from '@expressive-code/plugin-frames` is in the entry point module itself.  Something to consider to eliminate this would be to change the way the styling system works, possibly a registration API vs. the current approach.

Given the goal of this PR is to be 100% backwards compatible, in order to eliminate `plugin-frames` and `plugin-text-markers` "leaking" in to `Core`, I had to separate the `Core` related entry points into their own module.  Previously, all the `Core` related items were in the existing `index` entry point across `@expressive-code/plugin-shiki`, `expressive-code` and `rehype-expressive-code`.  I just added https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791 which is 100% functionality equivalent to the prior version of this PR but makes the following adjustments:

1. Moves `Core` related entries to [@expressive-code/plugin-shiki/core](https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791#diff-26082c3133f932d0384385536b31542e1b37c4520c3f0c5312595bc28071bfd9), [expressive-code/core](https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791#diff-0b9802efb37f68a99e078a883a64c710a23cf47edc6fa1aad607b2140a19c7d2) & [rehype-expressive-code/core](https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791#diff-b2bc80b05d2ecaff649198f47c84e6721a0d329cadd5e5cc336d00b296af0266)
2. Updates the [bundle-size.test.ts](https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791#diff-c27836fbe666b134979131fe3f1fff2d576e70424548e8e1346a95b9cb764fcc) to include tests around `plugin-frames` and `plugin-text-markers` for all 4 categories of tests within the suite

A few things:
1. I've added a few more comments to this PR for you to review based on things I found while working on https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791
2. The 'net' savings is only ~7k but it just didn't seem right to me to have `plugin-frames` & `plugin-text-markers` in the bundle when they were not being used at all.  This commit closes that gap. I also think that having the `core` entry points is more consistent with the `Core` concept that this PR is trying to introduce.
3. I've updated the description of this OP to reflect the final numbers and other information with this commit now included but here's a snapshot of how the numbers changed with https://github.com/expressive-code/expressive-code/pull/335/commits/ef112ae3d4844321af07ed72eaf472a484a6c791 - note that the actual v0.41.2 numbers themselves have changed from the original OP as I backported the tests to `v0.41.2` to get a 100% fully accurate comparison (previous v0.41.2 numbers were drawn from this PR vs. what's actually in `main`):

| 0 lang / 0 theme | Full Bundle | Shiki     | PluginShiki | PluginFrames | PluginTextMarkers |
| ---------------- | ----------- | --------- | ----------- | ------------ | ----------------- |
| v0.41.2          | ~~9,577,922~~ 9,575,587   | ~~9,695,299~~ 9,693,617  | ~~20,365~~ 19,415      | 23,186       | 22,246            |
| Core No Plugins  | ~~347,138~~ 338,581     | 0         | 0           | ~~3,788~~ 0            | ~~2,367~~ 0                 |
| Savings          | ~~9,230,784~~ 9,237,006   | ~~9,695,299~~ 9,693,617 | ~~20,365~~ 19,415      | ~~19,398~~ 23,186       | ~~19,879~~ 22,246            |

techfg commented on packages/expressive-code/src/index.ts:
```diff
@@ -1,16 +1,21 @@
 
+// NOTE - tsup code splitting is disabled to avoid re-exports being dropped
+// see https://github.com/evanw/esbuild/issues/1737 & https://github.com/evanw/esbuild/issues/1521
+// TODO: Consider a different bundler in order to support code-splitting to reduce /dist size (e.g., unbuild)
```
I had to disable splitting in `expressive-code` due to limitations with `tsup/esbuild` and `re-exports`.  I noticed that you also have it disabled in [@expressive-code/core](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/core/package.json#L38), likely for the same reason.

This isn't ideal but also not the end of the world since the total size of `dist` doesn't increase all that much but possibly it may be worth considering a bundler that has better support for `re-exports` such as [unbuild](https://github.com/unjs/unbuild) which would, more or less, be a drop-in replacement for tsup.

techfg commented on tsconfig.base.json:
```diff
@@ -15,14 +15,24 @@
     "target": "ESNext",
     "paths": {
       "@expressive-code/core": ["./packages/@expressive-code/core/src"],
+      "@expressive-code/core/hast": ["./packages/@expressive-code/core/src/hast"],
```
After adding `core` entry points to the three (3) packages, I started getting typescript warnings/errors due to unresolved packages.  I only needed to add entries here for the new `core` modules but noticed that several others were missing so added those in as well.  If their absence was intentional, let me know and I can remove all but the new `core` ones.

techfg commented on packages/rehype-expressive-code/src/common.ts:
```diff
@@ -0,0 +1,214 @@
+		baseStyles,
+		themeStyles,
+		jsModules,
+	}
```
As alluded to in https://github.com/expressive-code/expressive-code/pull/335#discussion_r2070204586, the more I've thought about this, I think that `ec` here should be typed `ExpressiveCodeEngine`.  If we make this change, then we do not need the [ctor](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66R16) option that gets passed into this function as we can always use `ExpressiveCodeCore` to create `ec` even when we're coming from an `ExpressiveCode` path.  If we make that change, then we could also eliminate the [loadTheme](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66R17) option and solely rely on [customLoadTheme](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b2bc80b05d2ecaff649198f47c84e6721a0d329cadd5e5cc336d00b296af0266R23).  To do that, in [createRenderer](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b5c8dbb131f4d919c674f58f434f328ebd2c4a88883bfa9a2b96e792181bffa0R23), we would adjust to set `customLoadTheme: options.customLoadTheme ?? loadShikiTheme`.

If we make both of those changes, we can significantly reduce the complexity of the types and "lift" everything in [common.ts](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66) directly in to [core.ts](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b2bc80b05d2ecaff649198f47c84e6721a0d329cadd5e5cc336d00b296af0266).

The only downsides to make these two changes are:
1. It would not be 100% backwards compatible with v0.41.2.  However, other than the `typings` themselves, at runtime, the only place I could see this causing any issues is if someone was doing `ec instanceof ExpressiveCode`, a situation that is highly unlikely to be done by userland.
2. We would be changing the actual options passed in by the user if they didn't provide a `customLoadTheme` so any downstream usage of options would now have a `customLoadTheme` when it originally didn't.  I do not think this is a problem though since if they are heading down the `createRenderer` path anyway, the only downstream usage of options is going to be stock functionality anyway.

techfg commented on packages/rehype-expressive-code/src/common.ts:
```diff
@@ -0,0 +1,214 @@
+	const jsModules = await ec.getJsModules()
+
+	return {
+		ec,
```
As alluded to in https://github.com/expressive-code/expressive-code/pull/335#discussion_r2070204586, the more I've thought about this, I think that `ec` here should be typed `ExpressiveCodeEngine`.  If we make this change, then we do not need the [ctor](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66R16) option that gets passed into this function as we can always use `ExpressiveCodeCore` to create `ec` even when we're coming from an `ExpressiveCode` path.  If we make that change, then we could also eliminate the [loadTheme](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66R17) option and solely rely on [customLoadTheme](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b2bc80b05d2ecaff649198f47c84e6721a0d329cadd5e5cc336d00b296af0266R23).  To do that, in [createRenderer](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b5c8dbb131f4d919c674f58f434f328ebd2c4a88883bfa9a2b96e792181bffa0R23), we would adjust to set `customLoadTheme: options.customLoadTheme ?? loadShikiTheme`.

If we make both of those changes, we can significantly reduce the complexity of the types and "lift" everything in [common.ts](https://github.com/expressive-code/expressive-code/pull/335/files#diff-4b727c4c354939182bb642b8f6a832edbb5d0bb67c35511f0fd079af1504ad66) directly in to [core.ts](https://github.com/expressive-code/expressive-code/pull/335/files#diff-b2bc80b05d2ecaff649198f47c84e6721a0d329cadd5e5cc336d00b296af0266).

The only downsides to make these two changes are:
1. It would not be 100% backwards compatible with v0.41.2.  However, other than the `typings` themselves, at runtime, the only place I could see this causing any issues is if someone was doing `ec instanceof ExpressiveCode`, a situation that is highly unlikely to be done by userland.
2. We would be changing the actual options passed in by the user if they didn't provide a `customLoadTheme` so any downstream usage of options would now have a `customLoadTheme` when it originally didn't.  I do not think this is a problem though since if they are heading down the `createRenderer` path anyway, the only downstream usage of options is going to be stock functionality anyway.

techfg reviewed (commented):
@hippotastic -

Found another opportunity to reduce bundle size as it relates to the default themes that are currently included in [ExpressiveCodeEngine](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/core/src/common/engine.ts#L237).

In short, when `themes` are provided in the config, the `github-dark` and `github-light` themes are still included in the bundle even though they will never be used.  This unnecessarily increases bundle size by 22,472 bytes.

To eliminate the increase when `themes` are provided, there are three options:

1. Eliminate these defaults from `ExpressiveCodeEngine` and move them to a higher level of abstraction (e.g., `ExpressiveCode` itself) and throw an error inside of ExpressiveCodeEngine if no themes have been specified (it would fail a little father down in the [code path](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/core/src/common/engine.ts#L384) anyway so an error gives something meaningful vs. an unhandled exception).  The only downside to this approach that I can see is that if anyone in userland is creating an instance of `ExpressiveCodeEngine` directly, this would be a breaking change.  My guess is it's unlikely anyone is creating a direct instance of the engine, however, and even in that case, it would be a simple adjustment to their code to include their desired themes.
2. Create a new `ExpressiveCodeEngineCore` and expose via a different entry point.  In short, `ExpressiveCodeEngine` would be a subclass of `ExpressiveCodeEngineCore` and layer in the default themes before calling `super`.  This approach would only be required to ensure 100% backwards compat but introduces an additional layer of complexity to the code base.  I've created an [example](https://github.com/techfg/expressive-code/commit/462d1192f0fd57f0eb0dbbf355486d0c76a9f1ef) of what this would look like.  Note that the name `@expressive-code/core/core` for the entry point likely needs a different name (e.g., possibly `base`).
3. A variant of option 2, this would instead be to create a different entry point (e.g., `full`) that contains everything in `@expressive-code/core` that isn't actually `Core` - for now, this would just be `ExpressiveCodeEngine`.  Given the name `core` is already in the package name and `Core` is implied, this make a little more sense than option 2 but it does mean `ExpressiveCodeEngine` would no longer be exported via `@expressive-code/core` and instead `@expressive-code/core/full` which would be a breaking change for anyone consuming `ExpressiveCodeEngine` directly.

At the end of the day, 22,472 bytes isn't the end of the world but it's still unnecessary in cases where userland wants full control of the bundle size by utilizing the `Core` layer.  I think option 1 is the most straightforward with only one likely minor drawback.  That said, option 2 & 3 provide some additional opportunities for potentially even further reducing bundle size for things that also exist in `@expressive-core/core` that may not be fully required for EC to function properly.  I haven't gone down the entire path yet of what is/isn't needed for baseline functionality, but this could further refine `@expressive-code/core` to truly be a `Core` layer and only contain the bare minimum.  

Lemme know your thoughts, thanks!
